onnxscript/version_converter/version_converter.py

```diff
@@ -1,55 +1,427 @@
+    "Size",
+    "Squeeze",
+    "Transpose",
+    "Unsqueeze",
+}
+
+
+_ADAPTERS_21_22 = {
+    "EyeLike",
+    "RandomUniform",
+    "RandomNormal",
+    "RandomUniformLike",
+    "RandomNormalLike",
+    "Multinomial",
+    "Bernoulli",
+    "ThresholdedRelu",
+    "Selu",
+    "Elu",
+    "Mish",
+    "HardSigmoid",
+    "HardSwish",
+    "Softsign",
+    "Softplus",
+    "Sin",
+    "Cos",
+    "Tan",
+    "Asin",
+    "Acos",
+    "Atan",
+    "Sinh",
+    "Cosh",
+    "Asinh",
+    "Acosh",
+    "Atanh",
+    "Round",
+    "Det",
+    "NegativeLogLikelihoodLoss",
+    "AveragePool",
+    "MaxPool",
+    "MaxUnpool",
+    "LpPool",
+    "MaxRoiPool",
+    "Conv",
+    "ConvTranspose",
+    "DeformConv",
+    "GlobalAveragePool",
+    "GlobalMaxPool",
+    "GlobalLpPool",
+    "InstanceNormalization",
+    "LpNormalization",
+    "Dropout",
+    "RoiAlign",
+    "RNN",
+    "GRU",
+    "LSTM",
+    "GridSample",
+}
+
+
+_ADAPTER_SETS = {
+    (18, 19): _ADAPTERS_18_19,
+    (19, 20): _ADAPTERS_19_20,
+    (20, 21): _ADAPTERS_20_21,
+    (21, 22): _ADAPTERS_21_22,
+}
+
+
+@dataclasses.dataclass
+class Replacement:
+    """A replacement for a node in the graph."""
+
+    new_outputs: Sequence[ir.Value]
+    new_nodes: Sequence[ir.Node]
+
+
+# A partial-evaluator function takes a node, a RewriterContext, OptimizerState and returns
+# a Replacement for the node or None (if no replacement is needed). It may also return just
+# the ir.Value or ir.Values to replace the output values of the node, when the new nodes
+# can be inferred from the RewriterContext used to build the new nodes.
+
+ReturnValue = Union[Replacement, Sequence[ir.Value], ir.Value, None]
+AdapterFunction = Callable[[ir.Node, orp.RewriterContext], ReturnValue]
+
+
+@dataclasses.dataclass
+class AdapterVersionChecker:
+    """A class that represents a version checker for a particular op.
+
+    It is applicable for a specific version upgrade (orignal_version -> target_version) of the op.
+    """
+
+    node_version: int | None
+    upgrade_version: int | None
+    function: AdapterFunction
+
+    def valid_for(self, opname: str, original_version: int, target_version: int) -> bool:
+        """Returns True if this evaluator is applicable for the given version upgrade."""
+        adapter_set = tuple((original_version, target_version))
+        if adapter_set not in _ADAPTER_SETS:
+            return False
+        if opname in _ADAPTER_SETS[adapter_set]:
+            return True
+        return False
+
+
+class AdapterRegistry:
+    """A class that maintains a registry of adapters for ops."""
+
+    def __init__(self):
+        self.op_adapters: dict[tuple[str, str], list[AdapterVersionChecker]] = {}
+
+    def lookup_adapters(
+        self, domain: str, opname: str, original_version: int, target_version: int
+    ):
+        adapter_list = self.op_adapters.get((domain, opname), [])
+        return [
+            adapter.function
+            for adapter in adapter_list
```

Comment:
I wonder if we really need to support a list of adapters (eg., for a given op from 18 to 19)? I understand that the constant-folding optimizer does this, but wondering if we need it here.